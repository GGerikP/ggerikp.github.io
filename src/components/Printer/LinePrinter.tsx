import React, { useCallback, useEffect, useRef, useState } from 'react';
import styled from 'styled-components';
import SegmentPrinter from './SegmentPrinter';

const PrintedLine = styled.div`
`;

interface LineSegment {
    text: string;
    link?: string;
    printDelayBefore?: number;
    printDelayAfter?: number;
}

export type Line = {
    lineSegments: LineSegment[];
}

type LinePrinterProps = {
    line: Line;
    lineIndex: number;
    typingSpeed: number;
    promptChars?: string;
    instantPrint: boolean;
    hasMoreLines: boolean;
    printNextLine?: () => void;
};

function LinePrinter({ line, lineIndex, typingSpeed, promptChars, instantPrint, hasMoreLines, printNextLine }: LinePrinterProps) {

    const sideEffectExecutedRef = useRef(false);
    const [segmentIndex, setSegmentIndex] = useState<number>(-1);
    const [displayedSegments, setDisplayedSegments] = useState<LineSegment[]>([]);

    // Called by the TextPrinter to print each next segment
    const printNextSegment = useCallback(() => {
        const updatedSegmentIndex = segmentIndex + 1;
        setSegmentIndex(updatedSegmentIndex);
        if (updatedSegmentIndex < line.lineSegments.length) {
            const nextSegment: LineSegment = {
                text: line?.lineSegments[updatedSegmentIndex]?.text,
                link: line?.lineSegments[updatedSegmentIndex]?.link,
                printDelayBefore: line?.lineSegments[updatedSegmentIndex]?.printDelayBefore,
                printDelayAfter: line?.lineSegments[updatedSegmentIndex]?.printDelayAfter,
            }
            const updatedDisplayedSegments = [...displayedSegments, nextSegment]
            setDisplayedSegments(updatedDisplayedSegments);
        } else {
            if (typeof printNextLine === 'function') {
                printNextLine();
            }
        }
    }, [segmentIndex, displayedSegments, line.lineSegments, printNextLine]);

    // Used to print the first segment
    useEffect(() => {
        if (!sideEffectExecutedRef.current) {
            printNextSegment();
        }
        sideEffectExecutedRef.current = true;
    }, [printNextSegment]);

    return (
        <PrintedLine id={`LinePrinter:${lineIndex}`}>
            {promptChars}
            {displayedSegments.map((segment, index) => {
                return (
                    <SegmentPrinter
                        key={index}
                        segment={segment}
                        segmentIndex={index}
                        typingSpeed={typingSpeed}
                        instantPrint={instantPrint}
                        hasMoreLines={hasMoreLines}
                        printNextSegment={printNextSegment}
                    />
                )
            })}
        </PrintedLine>
    );
}

export default LinePrinter;