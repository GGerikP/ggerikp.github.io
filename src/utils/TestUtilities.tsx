// Util Functions
export const checkIfCursorIsBlinking = (cursorElement: Element | null) => {
  const style = cursorElement ? getComputedStyle(cursorElement) : new CSSStyleDeclaration();
  expect(cursorElement).toBeInTheDocument();
  expect(cursorElement).toHaveStyle('display: inline-block;');
  expect(cursorElement).toHaveStyle('border-left: 2px solid white;');
  expect(style.getPropertyValue('animation')).toMatch(/.*\.5s step-end 0s infinite.*/);
};

export const checkIfCursorIsHidden = (cursorElement: Element | null) => {
  expect(cursorElement).toBeInTheDocument();
  expect(cursorElement).toHaveStyle('display: none;');
};

