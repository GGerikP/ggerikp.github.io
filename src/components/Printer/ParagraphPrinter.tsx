import React, { useCallback, useEffect, useState } from 'react';
import styled from 'styled-components';
import LinePrinter, { Line } from './LinePrinter';
import { CursorDisplay } from './Cursor';

const PrintedTextBlockContainer = styled.div`
    display: flex;
    flex-direction: column;
    flex-wrap: wrap;
    align-items: flex-start;
    text-align: left;
    width: 100%;
`;

type ParagraphPrinterProps = {
    id?: string;
    lines?: Line[];
    startingLineIndex?: number;
    typingSpeed: number;
    promptChars?: string;
    instantPrint: boolean;
    // eslint-disable-next-line no-unused-vars
    setPrintingState: (isPrinting: boolean) => void;
    finalCursorDisplay?: CursorDisplay;
};

function ParagraphPrinter ({
  id,
  lines,
  typingSpeed,
  promptChars,
  instantPrint,
  setPrintingState,
  finalCursorDisplay
}: ParagraphPrinterProps) {

  // const sideEffectExecutedRef = useRef(false);
  const [lineIndex, setLineIndex] = useState<number>(0);
  const [displayedLines, setDisplayedLines] = useState<Line[]>([]);

  const printNextLine = useCallback(() => {
    setLineIndex(prevLineIndex => prevLineIndex + 1);
  }, []);

  useEffect(() => {
    if (!lines || lineIndex >= lines?.length) {
      setPrintingState(false);
    } else {
      setPrintingState(true);
    }
  }, [lineIndex, lines, setPrintingState]);

  useEffect(() => {
    // console.log(`ParagraphPrinter: lineIndex(${lineIndex}), lines.length(${lines?.length}), displayedLines.length(${displayedLines.length})`);
    if (lines) {
      if (lineIndex < lines.length) {
        if (displayedLines[lineIndex] === undefined) {
          // console.log(`ParagraphPrinter: lineIndex(${lineIndex}) JSON.stringify(lines[lineIndex])(${JSON.stringify(lines[lineIndex])})`)
          if (!lines[lineIndex]) {
            // console.log(`Cannot add undefined line element at lines[${lineIndex}]`);
          } else {
            setDisplayedLines(prevDisplayedLines => {
              const updatedDisplayedLines = [...prevDisplayedLines];
              updatedDisplayedLines[lineIndex] = lines[lineIndex];
              return updatedDisplayedLines;
            });
          }
        } else {
          // console.log(`Not updating the displayedLines: displayedLines[${lineIndex}] is not undefined.`);
        }
      } else {
        // console.log(`Not updating the displayedLines: lineIndex is greater than or equal to the lines.length.`);
      }
    } else {
    // console.log('Not updating the displayedLines: there are no lines.');
    }
  }, [lineIndex, lines, displayedLines]);

  const paragraphId = (id ? id + '-' : '') + 'paragraph';
  return (
    <PrintedTextBlockContainer id={paragraphId}>
      {displayedLines && displayedLines.map((line, index) => {
        const LinePrinterID = paragraphId + `-line${index}`;
        return (
          <LinePrinter
            id={LinePrinterID}
            key={LinePrinterID}
            line={line}
            lineIndex={index}
            promptChars={promptChars}
            typingSpeed={typingSpeed}
            instantPrint={instantPrint}
            isLastLine={(!lines || index === lines.length - 1)}
            printNextLine={printNextLine}
            finalCursorDisplay={finalCursorDisplay}
          />
        );
      })}
    </PrintedTextBlockContainer>
  );
}

export default ParagraphPrinter;
