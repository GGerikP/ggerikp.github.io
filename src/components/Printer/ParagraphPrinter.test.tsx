import React from 'react';
import '@testing-library/jest-dom';
import ParagraphPrinter from './ParagraphPrinter';
import { Line } from './LinePrinter';
import { render, waitFor } from '@testing-library/react';
import { checkIfCursorIsBlinking, checkIfCursorIsHidden } from '../../utils/TestUtilities';

describe('ParagraphPrinter Unit Tests.', () => {
    it('Renders a ParagraphPrinter with 1 line with 1 segment', async () => {
        const lines: Line[] = [{
            segments: [{
                text: ''
            }]
        }]
        const renderedComponent = render(<ParagraphPrinter
            lines={lines}
            typingSpeed={0}
            promptChars='$ '
            instantPrint={false}
            setPrintingState={() => {}}
        />);
        expect(renderedComponent.asFragment()).toMatchSnapshot();
        expect(renderedComponent.container.querySelector('#ParagraphPrinter')).toBeInTheDocument();
        expect(renderedComponent.container.querySelector('#LinePrinter\\:0')).toBeInTheDocument();
        expect(renderedComponent.container.querySelector('#SegmentPrinter\\:0')).toBeInTheDocument();
        expect(renderedComponent.container.querySelector('#Cursor\\:0')).toBeInTheDocument();

    // Check the cursor - should be blinking
    checkIfCursorIsBlinking(renderedComponent.container.querySelector('#Cursor\\:0'));
  }, 500);

    it('Renders a ParagraphPrinter with 1 line with 2 segments', async () => {
        const lines: Line[] = [{
            segments: [{
                text: ''
            }, {
                text: ''
            }]
        }]
        const renderedComponent = render(<ParagraphPrinter
            lines={lines}
            typingSpeed={0}
            promptChars='$ '
            instantPrint={false}
            setPrintingState={() => {}}
        />);
        expect(renderedComponent.asFragment()).toMatchSnapshot();
        expect(renderedComponent.container.querySelector('#ParagraphPrinter')).toBeInTheDocument();
        expect(renderedComponent.container.querySelector('#LinePrinter\\:0')).toBeInTheDocument();
        waitFor(() => {
            expect(renderedComponent.container.querySelector('#SegmentPrinter\\:0')).toBeInTheDocument();
            expect(renderedComponent.container.querySelector('#Cursor\\:0')).toBeInTheDocument();
            checkIfCursorIsHidden(renderedComponent.container.querySelector('#Cursor\\:0'));

      expect(renderedComponent.container.querySelector('#SegmentPrinter\\:1')).toBeInTheDocument();
      expect(renderedComponent.container.querySelector('#Cursor\\:1')).toBeInTheDocument();
      checkIfCursorIsBlinking(renderedComponent.container.querySelector('#Cursor\\:1'));
    }, { timeout: 500 });
  }, 500);

    it('Renders a ParagraphPrinter with 5 lines, each with 5 segments', async () => {
        const lines: Line[] = [{
            segments: [{text: ''}, {text: ''}, {text: ''}, {text: ''}, {text: ''}]
        }, {
            segments: [{text: ''}, {text: ''}, {text: ''}, {text: ''}, {text: ''}]
        }, {
            segments: [{text: ''}, {text: ''}, {text: ''}, {text: ''}, {text: ''}]
        }, {
            segments: [{text: ''}, {text: ''}, {text: ''}, {text: ''}, {text: ''}]
        }, {
            segments: [{text: ''}, {text: ''}, {text: ''}, {text: ''}, {text: ''}]
        }]
        const renderedComponent = render(<ParagraphPrinter
            lines={lines}
            typingSpeed={0}
            promptChars='$ '
            instantPrint={true}
            setPrintingState={() => {}}
        />);
        expect(renderedComponent.asFragment()).toMatchSnapshot();
        expect(renderedComponent.container.querySelector('#ParagraphPrinter')).toBeInTheDocument();
        waitFor(() => {
            expect(renderedComponent.container.querySelector('#LinePrinter\\:0')).toBeInTheDocument();
            expect(renderedComponent.container.querySelector('#SegmentPrinter\\:0')).toBeInTheDocument();
            expect(renderedComponent.container.querySelector('#Cursor\\:0')).toBeInTheDocument();
            checkIfCursorIsHidden(renderedComponent.container.querySelector('#Cursor\\:0'));

      expect(renderedComponent.container.querySelector('#SegmentPrinter\\:1')).toBeInTheDocument();
      expect(renderedComponent.container.querySelector('#Cursor\\:1')).toBeInTheDocument();
      checkIfCursorIsHidden(renderedComponent.container.querySelector('#Cursor\\:1'));
    }, { timeout: 500 });
    waitFor(() => {
      expect(renderedComponent.container.querySelector('#LinePrinter\\:5')).toBeInTheDocument();
      expect(renderedComponent.container.querySelector('#SegmentPrinter\\:5')).toBeInTheDocument();
      expect(renderedComponent.container.querySelector('#Cursor\\:5')).toBeInTheDocument();
      checkIfCursorIsHidden(renderedComponent.container.querySelector('#Cursor\\:0'));

      expect(renderedComponent.container.querySelector('#LinePrinter\\:5 #SegmentPrinter\\:5')).toBeInTheDocument();
      expect(renderedComponent.container.querySelector('#LinePrinter\\:5 #SegmentPrinter\\:5 #Cursor\\:5')).toBeInTheDocument();
      checkIfCursorIsBlinking(renderedComponent.container.querySelector('#LinePrinter\\:5 #SegmentPrinter\\:5 #Cursor\\:5'));
    }, { timeout: 500 });
  }, 500);
});
