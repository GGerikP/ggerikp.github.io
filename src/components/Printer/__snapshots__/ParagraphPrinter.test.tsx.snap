// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParagraphPrinter Unit Tests. Renders a ParagraphPrinter with 1 line with 1 segment 1`] = `
<DocumentFragment>
  <div
    class="sc-jTQCzO gnzQdg"
    id="Tab0-paragraph"
  >
    <div
      class="sc-eDLKkx cfncZL"
      id="Tab0-paragraph-line0"
    >
      <span>
        $ 
      </span>
      <span
        class="sc-guDLey wuijx"
        id="Tab0-paragraph-line0-segment0"
      >
        <span
          class="sc-hLQSwg gcsUJF"
        />
        <div
          class="sc-beySPh brsoYh"
          id="Tab0-paragraph-line0-segment0-cursor"
        >
           
        </div>
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ParagraphPrinter Unit Tests. Renders a ParagraphPrinter with 1 line with 2 segments 1`] = `
<DocumentFragment>
  <div
    class="sc-jTQCzO gnzQdg"
    id="Tab0-paragraph"
  >
    <div
      class="sc-eDLKkx cfncZL"
      id="Tab0-paragraph-line0"
    >
      <span>
        $ 
      </span>
      <span
        class="sc-guDLey wuijx"
        id="Tab0-paragraph-line0-segment0"
      >
        <span
          class="sc-hLQSwg gcsUJF"
        />
        <div
          class="sc-beySPh brsoYh"
          id="Tab0-paragraph-line0-segment0-cursor"
        >
           
        </div>
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ParagraphPrinter Unit Tests. Renders a ParagraphPrinter with 5 lines, each with 5 segments 1`] = `
<DocumentFragment>
  <div
    class="sc-jTQCzO gnzQdg"
    id="Tab0-paragraph"
  >
    <div
      class="sc-eDLKkx cfncZL"
      id="Tab0-paragraph-line0"
    >
      <span>
        $ 
      </span>
      <span
        class="sc-guDLey wuijx"
        id="Tab0-paragraph-line0-segment0"
      >
        <span
          class="sc-hLQSwg gcsUJF"
        />
        <div
          class="sc-beySPh brsoYh"
          id="Tab0-paragraph-line0-segment0-cursor"
        >
           
        </div>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
